tests/phpunit/CRM/Mailing/BaseMailingSystemTest.php

```diff
@@ -0,0 +1,407 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | CiviCRM version 4.7                                                |
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC (c) 2004-2016                                |
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * Test that content produced by CiviMail looks the way it's expected.
+ *
+ * @package CiviCRM_APIv3
+ * @subpackage API_Job
+ *
+ * @copyright CiviCRM LLC (c) 2004-2016
+ * @version $Id: Job.php 30879 2010-11-22 15:45:55Z shot $
+ *
+ */
+
+/**
+ * Class CRM_Mailing_MailingSystemTest
+ * @group headless
+ * @see \Civi\FlexMailer\FlexMailerSystemTest
+ * @see CRM_Mailing_MailingSystemTest
+ */
+abstract class CRM_Mailing_BaseMailingSystemTest extends CiviUnitTestCase {
```

Comment:
Let's try this question again:

> why this is an abstract class .. I see no need for that and also u don't have any abstract method in it

From a design perspective, the answer is: because `BaseMailingSystemTest` should not be instantiated by itself. The content within the file `BaseMailingSystemTest.php` is a collection of general test scenarios which should be independent of particular implementations (e.g. current CiviMail vs FlexMailer). It would be meaningless to execute those general test scenarios unless you provide more details about the specific implementation to test.

> so you are defining it as abstract as a kind of protection from execution these test methods in the abstract class .. but I am looking at this class again and I think that you might do not need to add the word test before the method names .. so in this case you don't need to define the class as abstract.

From a functional perspective, it is desirable behavior that `phpunit` does not try to execute the `BaseMailingSystemTest` content by itself. That behavior could be achieved in multiple ways, but I don't see a point exploring those hypotheticals when there's no specific problem with the current concrete solution. FWIW, putting `abstract` on the class seems both simple and semantically accurate to me.